ext/compressors/lz4/lz4_compress.c

```diff
@@ -121,40 +138,143 @@ lz4_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
     size_t *result_lenp)
 {
 	WT_EXTENSION_API *wt_api;
-	char *compressed_data;
+	LZ4_PREFIX prefix;
 	int decoded;
-	size_t src_data_len;
+	uint8_t *dst_tmp;
+
+	(void)src_len;					/* Unused parameters */
 
 	wt_api = ((LZ4_COMPRESSOR *)compressor)->wt_api;
 
-	/* Retrieve compressed length from start of the data buffer. */
-	src_data_len = *(size_t *)src;
-	if (src_data_len + sizeof(size_t) > src_len) {
-		(void)wt_api->err_printf(wt_api,
-		    session,
-		    "lz4_decompress: stored size exceeds buffer size");
-		return (WT_ERROR);
+	/*
+	 * Retrieve the true length of the compressed block and source and the
+	 * decompressed bytes to return from the start of the source buffer.
+	 */
+	memcpy(&prefix, src, sizeof(LZ4_PREFIX));
+
+	/*
+	 * Decompress, starting after the prefix bytes. Use safe decompression:
+	 * we rely on decompression to detect corruption.
+	 *
+	 * Two code paths, one with and one without a bounce buffer. When doing
+	 * raw compression, we compress to a target size irrespective of row
+	 * boundaries, and return to our caller a "useful" compression length
+	 * based on the last complete row that was compressed. Our caller stores
+	 * that length, not the length of bytes actually compressed by LZ4. In
+	 * other words, our caller doesn't know how many bytes will result from
+	 * decompression, likely hasn't provided us a large enough buffer, and
+	 * we have to allocate a scratch buffer.
+	 */
```

Comment:
@Cyan4973, an additional issue I wanted to ask you about.

As you know, we currently have 3 values we're maintaining:
* the length of "useful" uncompressed bytes we want from the compressed chunk,
* the length of the compressed chunk,
* the length of the uncompressed chunk.

Because we don't know the number of compressed bytes associated with the useful uncompressed bytes we want, we have to give `LZ4_decompress_safe` a buffer big enough to hold the complete uncompressed chunk, which, for layering reasons, is difficult for us: we're discussing alternatives, but at the moment we're allocating a scratch buffer and doing a memory copy, which will slow things down.

@michaelcahill, raised a point you and I briefly touched on: right now, `LZ4_decompress_safe ` returns an error if the number of compressed bytes is non-zero when we've run out of buffer into which we're decompressing.

Would it be possible for `LZ4_decompress_safe` to allow that usage?

In other words, is there some way to tell `LZ4_decompress_safe` we only want some number of uncompressed bytes from the compressed source, and if we fill the buffer before the number of compressed source bytes goes to 0, that's OK?

Might this be useful to other applications, too? I could imagine an application that only needs the start of a compressed block.

Alternatively, if there's a specific error guaranteed to be returned in this case, we could treat that error as success?